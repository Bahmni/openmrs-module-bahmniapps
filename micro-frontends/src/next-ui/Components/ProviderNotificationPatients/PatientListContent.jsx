import React, { useState } from 'react';
import { Button, TextArea } from "carbon-components-react";
import { FormattedMessage } from 'react-intl';
import "./PatientListContent.scss";
import { Title } from "bahmni-carbon-ui";
import { formatArrayDateToDefaultDateFormat } from "../../utils/utils";
import PropTypes from "prop-types";

const PatientListContent = ({ patientMedicationDetails, handleOnClick }) => {
  const { administered_date_time, administered_drug_name, medication_administration_performer_uuid, medication_administration_uuid } = patientMedicationDetails;
  const acknowledgementRequiredText = (<FormattedMessage id="AKNOWLEDGE_BUTTON" defaultMessage="Acknowledge" />);
  const [providerNotes, setProviderNotes] = useState("");

  function acknowledgeMedication() {
    handleOnClick(medication_administration_performer_uuid, medication_administration_uuid, providerNotes);
    setProviderNotes("");
  }

  const handleNotesChange = (e) => {
    setProviderNotes(e.target.value);
  };

  return (
    <div className="patient-list-content">
      <span>{formatArrayDateToDefaultDateFormat(administered_date_time)}</span>
      <div className="content-info">
        <span>{administered_drug_name}</span>
        <div className="notes">
          <TextArea
            className="patient-list-text-area"
            labelText={<Title text="Note" isRequired />}
            placeholder="Enter Notes"
            rows={1}
            required
            value={providerNotes}
            onChange={handleNotesChange}
          />
          <Button
            className="patient-list-button"
            disabled={providerNotes.trim() === ""}
            onClick={acknowledgeMedication}
          >
            {acknowledgementRequiredText}
          </Button>
        </div>
      </div>
    </div>
  );
};

PatientListContent.propTypes = {
    patientMedicationDetails: PropTypes.shape({
        administered_date_time: PropTypes.array.isRequired,
        administered_drug_name: PropTypes.string.isRequired,
        medication_administration_performer_uuid: PropTypes.string.isRequired,
    }).isRequired,
    handleOnClick: PropTypes.func.isRequired
};

export default PatientListContent;
