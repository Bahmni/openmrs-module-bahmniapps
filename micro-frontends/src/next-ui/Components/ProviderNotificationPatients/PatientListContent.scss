.patient-list-content {
    display: flex;
    flex-flow: row nowrap;
    justify-content: start;
    padding: 10px;

    .content-info {
        margin-left: 20px;
        display: flex;
        flex-flow: column nowrap;
        justify-content: start;
        align-items: start;

        .notes {
            display: flex;
            flex-flow: row wrap;
            justify-content: start;
            align-items: end;

            .patient-list-text-area {
                flex: 1;
                min-height: 48px;
            }

            .patient-list-button {
                width: 116px;
                height: 48px;
                margin-left: 10px;

                @media screen and (max-width: 640px) {
                    margin-left: 0;
                    margin-top: 10px;
                }
            }
        }
    }
}
