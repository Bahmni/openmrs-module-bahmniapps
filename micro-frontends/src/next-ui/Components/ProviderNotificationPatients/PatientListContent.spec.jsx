import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PatientListContent from './PatientListContent';

describe('PatientListContent Component', () => {
  const patientMedicationDetails = {
    administered_date_time: [2022,6,9],
    administered_drug_name: 'Drug A',
    medication_administration_performer_uuid: 'performer123',
    medication_administration_uuid: 'medication123'
  };

  const handleOnClickMock = jest.fn();

  it('renders the component without crashing', () => {
    const {container} = render(
        <PatientListContent
          patientMedicationDetails={patientMedicationDetails}
          handleOnClick={handleOnClickMock}
          
        />
      );
    expect(container).toMatchSnapshot();
  });

  it('displays the correct medication information', () => {
    const { getByText } = render(
      <PatientListContent
        patientMedicationDetails={patientMedicationDetails}
        handleOnClick={handleOnClickMock}
      />
    );
    expect(getByText('9/6/2022')).toBeTruthy();
    expect(getByText('Drug A')).toBeTruthy();
  });

  it('calls handleOnClick with correct arguments when Acknowledge button is clicked', () => {
    const { getByText, getByPlaceholderText } = render(
      <PatientListContent
        patientMedicationDetails={patientMedicationDetails}
        handleOnClick={handleOnClickMock}
      />
    );
    const notesInput = getByPlaceholderText('Enter Notes');
    fireEvent.change(notesInput, { target: { value: 'Some notes' } });
    expect(notesInput.value).toBe('Some notes');
    fireEvent.click(getByText('Acknowledge'));
    expect(handleOnClickMock).toHaveBeenCalledTimes(1);
    expect(handleOnClickMock).toHaveBeenCalledWith(
      'performer123',
      'medication123',
      'Some notes'
    );
  });

  it('updates providerNotes state when notes are entered', () => {
    const { getByPlaceholderText } = render(
      <PatientListContent
        patientMedicationDetails={patientMedicationDetails}
        handleOnClick={handleOnClickMock}
      />
    );
    const notesInput = getByPlaceholderText('Enter Notes');
    fireEvent.change(notesInput, { target: { value: 'Some notes' } });
    expect(notesInput.value).toBe('Some notes');
  });

  it('clears providerNotes state after Acknowledge button is clicked', () => {
    const { getByPlaceholderText, getByText } = render(
      <PatientListContent
        patientMedicationDetails={patientMedicationDetails}
        handleOnClick={handleOnClickMock}
      />
    );
    const notesInput = getByPlaceholderText('Enter Notes');
    fireEvent.change(notesInput, { target: { value: 'Some notes' } });
    expect(notesInput.value).toBe('Some notes');
    fireEvent.click(getByText('Acknowledge'));
    expect(handleOnClickMock).toHaveBeenCalledWith(
      'performer123',
      'medication123',
      'Some notes'
    );
    expect(notesInput.value).toBe('');
  });

  it('disables Acknowledge button when no notes are entered', () => {
    const { getByText } = render(
      <PatientListContent
        patientMedicationDetails={patientMedicationDetails}
        handleOnClick={handleOnClickMock}
      />
    );
    const acknowledgeButton = getByText('Acknowledge');
    expect(acknowledgeButton).toHaveProperty('disabled', true); 
  });

  it('enables Acknowledge button when notes are entered', () => {
    const { getByText, getByPlaceholderText } = render(
      <PatientListContent
        patientMedicationDetails={patientMedicationDetails}
        handleOnClick={handleOnClickMock}
      />
    );
    const notesInput = getByPlaceholderText('Enter Notes');
    fireEvent.change(notesInput, { target: { value: 'Some notes' } });
    const acknowledgeButton = getByText('Acknowledge');
    expect(acknowledgeButton).toHaveProperty('disabled', false); 
  });
});
