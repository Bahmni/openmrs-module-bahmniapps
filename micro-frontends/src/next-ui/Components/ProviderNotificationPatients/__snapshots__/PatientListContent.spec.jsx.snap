// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatientListContent Component renders the component without crashing 1`] = `
<div>
  <div
    class="patient-list-content"
  >
    <span>
      9/6/2022
    </span>
    <div
      class="content-info"
    >
      <span>
        Drug A
      </span>
      <div
        class="notes"
      >
        <div
          class="bx--form-item"
        >
          <div
            class="bx--text-area__label-wrapper"
          >
            <label
              class="bx--label"
            >
              <div
                class="titleText"
              >
                <span>
                  Note 
                </span>
                <span
                  class="required"
                >
                  *
                </span>
              </div>
            </label>
          </div>
          <div
            class="bx--text-area__wrapper"
          >
            <textarea
              class="bx--text-area patient-list-text-area"
              cols="50"
              placeholder="Enter Notes"
              required=""
              rows="1"
            />
          </div>
        </div>
        <button
          class="patient-list-button bx--btn bx--btn--primary bx--btn--disabled"
          disabled=""
          tabindex="0"
          type="button"
        >
          Acknowledge
        </button>
      </div>
    </div>
  </div>
</div>
`;
